cmd/skopeo/proxy.go

```diff
@@ -0,0 +1,265 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/containers/image/v5/image"
+	"github.com/containers/image/v5/manifest"
+	"github.com/containers/image/v5/pkg/blobinfocache"
+	"github.com/containers/image/v5/transports/alltransports"
+	"github.com/containers/image/v5/types"
+	"github.com/opencontainers/go-digest"
+	"github.com/spf13/cobra"
+)
+
+type proxyHandler struct {
+	imageref string
+	sysctx   *types.SystemContext
+	cache    types.BlobInfoCache
+	imgsrc   *types.ImageSource
+	img      *types.Image
+	shutdown bool
+}
+
+func (h *proxyHandler) ensureImage() error {
+	if h.img != nil {
+		return nil
+	}
+	imgRef, err := alltransports.ParseImageName(h.imageref)
+	if err != nil {
+		return err
+	}
+	imgsrc, err := imgRef.NewImageSource(context.Background(), h.sysctx)
+	if err != nil {
+		return err
+	}
+	img, err := image.FromUnparsedImage(context.Background(), h.sysctx, image.UnparsedInstance(imgsrc, nil))
+	if err != nil {
+		return fmt.Errorf("failed to load image: %w", err)
+	}
+	h.img = &img
+	h.imgsrc = &imgsrc
+	return nil
+}
+
+func (h *proxyHandler) implManifest(w http.ResponseWriter, r *http.Request) error {
+	if err := h.ensureImage(); err != nil {
+		return err
+	}
+
+	_, err := io.Copy(io.Discard, r.Body)
+	if err != nil {
+		return err
+	}
+	ctx := context.TODO()
+	rawManifest, _, err := (*h.img).Manifest(ctx)
+	if err != nil {
+		return err
+	}
+	digest, err := manifest.Digest(rawManifest)
+	if err != nil {
+		return err
+	}
+	w.Header().Add("Manifest-Digest", digest.String())
+
+	ociManifest, err := manifest.OCI1FromManifest(rawManifest)
```

Comment:
Alternatively, if this doesn’t specifically need an OCI-formatted manifest, `types.Image.Inspect` provides some information from an image of any format; in particular that’s a reasonable way to list all layer digests, enabling use of `/blobs/`, if that were the only thing necessary.

Reply:
I’ll read the linked proposal (and the updated code) later, for now just a very quick first impression:

- I’m _very_ reluctant to turn all of `ImageDestination` into a stable RPC API. We are actually already starting to hide parts of it from the public type (compare `c/image/internal/types`) just because we don’t know how it is going to evolve. Sure, API versioning and capability checks and…, but that’s an extra maintenance that’s basically known to materialize at least once this year again.
- Could you describe what are you building, at a … higher but still somewhat low :) … level, please?
 
  I kinda thought, to be honest without investigating at all, that the “encapsulate host operating system updates” implies specially-built images only usable with a specialized consumer, so it should be possible to to hard-code the assumption that the images are in the OCI (or schema2) format in the producer of the image as well as the consumer. So I was expecting this to just refuse non-OCI, or to be a dumb wrapper around `ImageSource.GetManifest` returning the MIME type + byte array as is (and for the client to then refuse non-OCI).

  Alternatively, I was thinking that the application doesn’t care about the internals of the manifest format and can just use `types.Image.Inspect` to extract the format-independent aspects of the manifest/config, like the layer list.

  If neither is true, there are probably some extra requirements I’m not thinking about.

  (To be clear about the cost, if this is supposed to accept schema1 images and always provide an OCI-formatted image, the server will need, for the conversion, to read _all_ of the layers _and decompress them_, and then keep a state of the generated config+manifest. Compare the earlier discussions WRT consuming a “streaming OCI archive” destination — the data must be buffered at _some_ point because non-byte-for-byte-identical image copies can, in general, only produce the manifest last, but the consumers need the manifest first; that’s the case for compression/decompression of the image, as well as format conversions, at least if schema1 is involved.)

Reply:
OK, I pushed an update which now accepts only oci and schema2 (converted to oci) and rejects everything else.